src/map/storage.c

```diff
@@ -212,21 +231,29 @@ int storage_additem(struct map_session_data* sd, struct item* item_data, int amo
  *------------------------------------------*/
 int storage_delitem(struct map_session_data* sd, int n, int amount)
 {
+	struct item *it = NULL;
+
 	nullpo_retr(1, sd);
-	Assert_retr(1, n >= 0 && n < MAX_STORAGE);
-	if( sd->status.storage.items[n].nameid == 0 || sd->status.storage.items[n].amount < amount )
-		return 1;
 
-	sd->status.storage.items[n].amount -= amount;
-	if( sd->status.storage.items[n].amount == 0 )
-	{
-		memset(&sd->status.storage.items[n],0,sizeof(sd->status.storage.items[0]));
-		sd->status.storage.storage_amount--;
-		if( sd->state.storage_flag == STORAGE_FLAG_NORMAL )
-			clif->updatestorageamount(sd, sd->status.storage.storage_amount, MAX_STORAGE);
+	Assert_retr(1, n >= 0 && n < VECTOR_LENGTH(sd->storage.item));
+
+	nullpo_retr(1, (it = &VECTOR_INDEX(sd->storage.item, n)));
+
+	Assert_retr(1, amount <= it->amount);
+
+	Assert_retr(1, it->nameid > 0);
+
+	it->amount -= amount;
+
+	if (it->amount == 0) {
+		memset(it, 0, sizeof(struct item));
+		clif->updatestorageamount(sd, --sd->storage.aggregate, MAX_STORAGE);
 	}
-	if( sd->state.storage_flag == STORAGE_FLAG_NORMAL )
-		clif->storageitemremoved(sd,n,amount);
+
+	sd->storage.save = true;
+
+	clif->storageitemremoved(sd, n, amount);
```

Comment:
In the case of `@clearstorage` or the itemcheck, you shouldn't send this packet (that's the reason for the `STORAGE_FLAG_NORMAL` check in the original code: if the client isn't showing the storage, we shouldn't send an update packet)